### ISSUE TITLE
Allow to pass `sample_weight` to estimators in ShapRFECV

### ISSUE THREAD

**Issue by timlod** [ing-bank/probatus#106]
**Problem Description**
Currently, there is no option to pass sample weights to the estimators being fit.

**Desired Outcome**
I would like to pass sample weightings to address class imbalance in a more goal-oriented way.

**Solution Outline**
Allow to pass sample weights into the `fit_compute` call.

**Comment by Matgrb** [comment]
I think this would be a great addition, and not only in `ShapRFECV`, but in all modules that perform model training. We can definitely start with that module. I fully support adding it to the probatus `fit` and `fit_compute` interface. 

For `ShapRFECV` implementing it is slightly more complex than the other modules, because we need to apply sample weights correctly,  while calculating SHAP values using Cross-validation. 

The main changes that need to be done are:
1. Add `sample_weights` as `fit` and `fit_compute` parameters and describe their docstrings in a way that is consistent with sklearn description of this parameters. This way the users seeing it,  will directly understand why this is useful.
2. Make a function to preprocess `sample_weights` and throw exceptions in case the format of it is wrong. The preprocessing should be done from list, np.array to pd.Series, with making sure that the index is the same as in X after applying `preprocess_data` function. This part is crucial, because then we can apply the same preprocessing in all modules.
3. Pass sample weights to `fit` method of the `clf` correctly.
4. Pass sample weights to `_get_feature_shap_values_per_fold`. Within that function the sample weights need to be passed to `clf.fit`, `scorer` and `shap_calc`
5. Pass sample weights to `calculate_shap_importance` to take them into consideration while calculating feature importance from cross-validated results.
6. Make unit tests for checking if passing sample_weights to every function that is modified works as expected. The test might be super simple, and such that you can fully control the output,  e.g. all weights to 0 and only a couple of them to 1.

Would you like to pick up this issue?  I can assist you in implementation. If you prefer, I can also pick this issue up, just to make it easier to implement sample weights in other modules of probatus.

Side note:
One very easy way to ensure class balance would also be passing `class_weights="balanced"` at the initialization of the classifier. This is a bit more limited solution, since it does not allow you to specify per sample different importance, however, it is available right away in case you need it now.

**Comment by timlod** [comment]
I'm trialling probatus as part of a larger piece of work that's taking all of my time currently - I may be able to pick it up after that is done (~2 weeks from now approx.).
I'm aware of `class_weights="balanced"` - I have other information that leads to more appropriate weighting which I am using across all models so far.

**Comment by Matgrb** [comment]
Great to hear! Then let us know in a bit if you want to pick that one up and if you have any questions regarding the implementaton.

**Comment by timlod** [comment]
Hi, I am working on this now, after some delays.

Some things to consider:
1. I just merged from master and noticed the EarlyStopping work. In there, it uses both training and validation set for validation (line 913) - I think it should be just the second tuple?
2. I am not sure we should use `sample_weight` also for the scorers - e.g., in my specific problem, I am using sample weights to do custom class imbalance handling, and I still want to score against an unweighted formulation on any validation set.

What are your thoughts on this?

**Comment by Matgrb** [comment]
Indeed i think you might be right, @operte could you double check this one? If lgbm uses first eval metric for early stopping, this way will never stop, because train score will always increase :) Good catch!

Second point is quite tricky. If we only allow for training with sample weights, but not evaluating, some people might prefer to do both. I personally think we should do both, because weights might also define how important samples are in your test set, or in cross-validation. It could also depend on the use case. Implementing both options will add a lot of complexity to the already complex code, I would rather implement one of them. Maybe the solution would be to see how sample weights are used within cross validation in sklearn. This way we are consistent with that package. What do you think?

Indeed the categorical features are an issue now, could you make an issue about it with the description you have added here?

**Comment by timlod** [comment]
I will add the issue about categorical features - I can also make a PR - I fixed all of this on my clone.

In sklearn, it really depends on how you implement cross validation. For example, https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html#sklearn.model_selection.cross_validate does not accept `sample_weight`. Mainly the sample weights are used in the fit calls, and you have to explicitly pass them to every step where you want them. https://github.com/scikit-learn/scikit-learn/issues/4497 shows that the treatment thereof has been a long standing issue - they're just not used very consistently throughout.
Personally, I wouldn't want to make the PR including the use of SW for evaluation - simply because that's not the behaviour I need and I'm using them right now only for fitting. So I'd advocate for using them only for fitting, or try to find a way to do both without raising complexity too much (a single flag could do this - would that be too much?). In any case, this needs to be well-documented (eval metrics can change drastically depending on your sample weights).

**Comment by Matgrb** [comment]
I see the point, indeed i think the sample_weights should mainly be passed to fit method of probatus and further of the model.
We need top give an informative description of that, and possibly
```
if sample_weights not None and verbose>0:
     warn("Sample weights are passed only to fit method of the model.") # or something like that.
```

Good also to make a very short notebook in howto section of the documentation, on how to use sample weights in ShapRFECV.

**Comment by timlod** [comment]
About EarlyStopping - may I ask why this is not implemented as a parameter to `ShapRFECV`? Currently I'm implementing `sample_weight` in both classes, but I feel that a parameter `early_stopping` in `ShapRFECV` would do just as well, minimising code duplication.

**Comment by Matgrb** [comment]
#39 Lengthy answer is here.

In a nutshell:
- Early stopping applies only to 2-3 boosting models, the rest of the models we support in probatus e.g. linear, other tree-based models do not apply it. Early stopping requires adding 2 more parameters, and since it has a very narrow scope, we don't want to increase complexity of ShapRFECV API
- Early stopping on its own is a kind of hyperparam optimization. So in this case, it doesn't make sense to optimize the model at each step, because the training might be stopped before n_trees is reached. You can simply use lgbm with 500 small trees, and let it stop at the right moment. So we split the functionality into two: Hyperparameter opt and longer run time in ShapRFECV, Early stopping and no extra hyperparam opt, for very large datasets for speed.

**Comment by operte** [comment]
Regarding the `eval_set`, I think you are right, we should only have the validation set there. We misinterpeted this in the past. For some reason I was convinced that you could have several `eval_set`s but only the last one would be considered for early stopping. Re-reading [this](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.early_stopping.html) it does not sound to be the case. 

I will pick this up :)

**Comment by timlod** [comment]
My code is pretty much ready, but I notice that `warnings.warn` is never printed, even with verbosity > 100.
I added this to check and don't get the warning:
```
if self.verbose > 0:
    print("no warning?")
    warnings.warn(
         "sample_weight is passed only to the fit method of the model, not the evaluation metrics."
     )
```

Note: I'm calling everything from a jupyter kernel.

@operte I'm also fixing this inside the PR I'm submitting, because I wanted to test it with the proper implementation of `EarlyStoppingShapRFECV`. Haven't looked at any potential tests for this though (and don't have time to do more than removing the eval_set right now.

**Comment by Matgrb** [comment]
You can make a PR, and I will review the code.

For the issue with warnings, i will also check it today if they show up for me.

**Comment by timlod** [comment]
Fixed in #139.
